src/common/src/common-json.c

```diff
@@ -0,0 +1,766 @@
+   BSON_UNUSED (iter);
+   BSON_UNUSED (key);
+   if (mcommon_json_append_bson_array (state->append, v_array, state->mode, state->max_depth)) {
+      return !mcommon_string_append_status (state->append);
+   } else {
+      state->is_corrupt = true;
+      return true;
+   }
+}
+
+bool
+mcommon_json_append_bson_values (
+   mcommon_string_append_t *append, const bson_t *bson, bson_json_mode_t mode, bool has_keys, unsigned max_depth)
+{
+   mcommon_json_append_visit_t state = {.append = append, .max_depth = max_depth, .mode = mode, .has_keys = has_keys};
+   bson_iter_t iter;
+   if (!bson_iter_init (&iter, bson)) {
+      return false;
+   }
+   static const bson_visitor_t visitors = {
+      mcommon_json_append_visit_before,     mcommon_json_append_visit_after,     mcommon_json_append_visit_corrupt,
+      mcommon_json_append_visit_double,     mcommon_json_append_visit_utf8,      mcommon_json_append_visit_document,
+      mcommon_json_append_visit_array,      mcommon_json_append_visit_binary,    mcommon_json_append_visit_undefined,
+      mcommon_json_append_visit_oid,        mcommon_json_append_visit_bool,      mcommon_json_append_visit_date_time,
+      mcommon_json_append_visit_null,       mcommon_json_append_visit_regex,     mcommon_json_append_visit_dbpointer,
+      mcommon_json_append_visit_code,       mcommon_json_append_visit_symbol,    mcommon_json_append_visit_codewscope,
+      mcommon_json_append_visit_int32,      mcommon_json_append_visit_timestamp, mcommon_json_append_visit_int64,
+      mcommon_json_append_visit_maxkey,     mcommon_json_append_visit_minkey,    NULL, /* visit_unsupported_type */
+      mcommon_json_append_visit_decimal128,
+   };
+   /* Note that early exit from bson_iter_visit_all does not affect our success, which is based only on BSON validity.
+    * BSON errors will set is_corrupt if they prevent full traversal, but non-fatal parse errors (like invalid UTF-8)
+    * may let bson_iter_visit_all() succeed while leaving an error status in iter.err_off. */
+   (void) bson_iter_visit_all (&iter, &visitors, &state);
+   return iter.err_off == 0 && !state.is_corrupt;
+}
+
+static BSON_INLINE bool
+mcommon_json_append_bson_container (mcommon_string_append_t *append,
+                                    const bson_t *bson,
+                                    bson_json_mode_t mode,
+                                    unsigned max_depth,
+                                    bool has_keys,
+                                    const char *empty,
+                                    const char *begin_non_empty,
+                                    const char *end_non_empty,
+                                    const char *omitted)
+{
+   // Note that the return value here is bson validity, not append status.
+   if (bson_empty (bson)) {
+      (void) mcommon_string_append (append, empty);
+      return true;
+   } else if (max_depth == 0) {
+      (void) mcommon_string_append (append, omitted);
```

Comment:
Is `max_depth` expected to be 0? If no, consider using `BSON_ASSERT (max_depth > 0);` to remove an unused special case.